### ISSUE TITLE
Feature request: generic layer-window hook - run layers [il_start, il_end) with an injectable/extractable boundary residual

### ISSUE THREAD

**Issue by PavanManchikatla** [ggml-org/llama.cpp#25577]
### Prerequisites
- [x] Running latest code (checkout `13f2b28b`).
- [x] Searched open/closed issues (found related #22436, #23568, different ask, see below) and Discussions (#20252 pipeline parallelism).
- [x] This is a new, useful enhancement.

## Feature description

I'd like a small, architecture-agnostic option to run only transformer layers `[il_start, il_end)`. A non-first shard consumes an injected hidden state as the layer-`il_start` input. A non-last shard emits the raw residual entering layer `il_end` instead of logits.

```c
// llama_context_params (or a setter)
int32_t il_start;   // default 0
int32_t il_end;     // default -1 => n_layer
```

I'd want this honored by the shared graph driver, so each arch's builder doesn't have to re-implement it.

## Motivation

I hit this while splitting a model across machines. Cross-machine pipeline (layer) sharding runs a contiguous layer range on each node and passes the boundary residual to the next, so it needs exactly this primitive. It's the building block under the end-user features requested in #22436 (pipeline parallelism over TCP/IP) and #23568 (TP+PP). Today there's no supported way to run a layer subrange starting from an injected residual and read out a boundary residual, so every such effort re-patches the graph builders.

Two of the three pieces already exist and work well:

- Inject. `llama_batch.embd` feeds an external `[n_embd, n_tokens]` residual through `build_inp_embd` (unscaled).
- Per-layer residual capture. `llm_graph_result::t_layer_inp[il]` + `get_layer_inp(il)` and the `embeddings_layer_inp` machinery already record each layer's input hidden state.

The only missing primitive is restricting the layer loop to a range (and, for a boundary shard, skipping the final norm + lm_head and surfacing the layer-`il_end` input residual).

## Supporting evidence: two bugs I found while prototyping this

While implementing a stopgap per-arch patch (below), two issues surfaced that anyone doing partial-graph execution will hit.

1. Embeddings-context defaults to non-causal attention. A context reused to extract a mid-model residual must `llama_set_causal_attn(ctx, true)`, or the residual diverges from the causal model. Easy to miss, worth documenting or defaulting per arch.
2. `inp_out_ids` dangles on a non-final shard. The output-id `get_rows` is keyed on `il == n_layer - 1`. When a shard's last built layer is `< n_layer-1`, `inp_out_ids` is never consumed, and `llm_graph_input_out_ids::set_input` then dereferences its null buffer (`ggml_backend_buffer_is_host(NULL)`, assert). Keying it on the shard's actual last layer fixes it.

I'm happy to open these as separate issues if you'd prefer to track them independently.

## Stopgap I use today

A ~47-line patch adds `il_start/il_end` to `llama_context_params`/`llama_cparams`, loops `[l0,l1)`, keys the `inp_out_ids` `get_rows` on the shard's last layer, and for a boundary shard exposes the pre-norm residual. I verified it bit-exact vs the unsplit model (F32 boundary, CPU and Metal). It's per-arch (`llama.cpp`, `qwen2.cpp`) only because each dense arch has its own builder in `src/models/*.cpp`, which is the reason I'm asking for a generic hook. I'm happy to upstream a generic version.

## Why not RPC / `--split-mode layer`?

Those distribute one logical graph across local devices. They don't give independent per-shard execution with an injectable boundary residual and logits-from-injected-hidden-state, which is what cross-machine pipeline sharding needs.

---
Context: this came up while I was building [Hydra](https://github.com/PavanManchikatla/Hydra), a crash-safe LAN pipeline-inference runtime.

**Comment by PavanManchikatla** [comment]
Gentle bump — since filing, we've built a working proof of the layer-window API (a ~47-line per-arch patch on `llama`/`qwen2`) and used it to pipeline-shard a model across 3 heterogeneous machines with crash-safe recovery. Happy to open the two prototyping bugs (non-causal embeddings default; dangling `inp_out_ids`) as separate issues if that's easier to triage. Is an enabling primitive like this something the project would consider?
